File: BaekJoon/solved/JS/5076 Web Pages.js
const fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const BOJ_5076 = () => {
    let i = 0;
    while ((S = input[i++]) != "#")
    {
        let stk = [];
        for (let i = 0, l = S.length; i < l; i++)
        {
            if (S[i] == '<')
            {
                let inner = S.slice(++i, S.indexOf('>', i));
                if (inner == "br /")
                {
                    i += 5;
                    continue;
                }
                let idx = inner.indexOf(' ')
                let tag = idx == -1 ? inner : inner.slice(0, idx);
                if (inner[0] == '/')
                {
                    let size = stk.length;
                    if (!size || stk[size-1] !== inner.slice(1))
                    {
                        stk.push(1);
                        break;
                    }
                    stk.pop();
                }
                else stk.push(tag);

                i += inner.length + 1;
            }
        }
        console.log(stk.length ? "illegal" : "legal");
    }
}

BOJ_5076();

File: BaekJoon/solved/C++/5076_Web_Pages.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

void BOJ_5076()
{
    while (1)
    {
        string S;
        getline(cin, S);
        if (S[0] == '#') break;

        stack<string> s;
        int b, i = 0;
        char c;
        while ((c = S[i++]))
        {
            if (c == '<')
            {
                // <br /> 건너 뛰기
                if (S[i] == 'b' && S[i + 1] == 'r')
                    i += 4;
                // 닫히는 태그
                else if (S[i] == '/')
                {
                    b = ++i;
                    while (S[i] != '>')
                        i++;
                    if (s.empty() || s.top() != S.substr(b, i - b))
                    {
                        s.push("x");
                        break;
                    }
                    else s.pop();
                }
                // 열리는 태그
                else
                {
                    b = i;
                    while (S[i] != '>' && S[i] != ' ')
                        i++;
                    s.push(S.substr(b, i - b));
                }
            }
        }
        puts(s.empty() ? "legal" : "illegal");
    }
}

int main()
{
    BOJ_5076();
    return (0);
}